File: Day097_2025/Q097_Leetcode.py
from typing import List
class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        total = sum(nums)
        if total % 2 != 0:
            return False
        target = total // 2

        prev = [False] * (target + 1)
        curr = [False] * (target + 1)
        prev[0] = True
        curr[0] = True

        for i in reversed(range(len(nums))):
            for t in range(target + 1):
                include = prev[t - nums[i]] if t - nums[i] >= 0 else False
                exclude = prev[t]
                curr[t] = include or exclude
            prev = curr[:]

        return prev[target]

File: Day097_2025/Q097_Leetcode.java
class Q097_Leetcode {
    public boolean canPartition(int[] nums) {
        int total = 0;
        for (int num : nums) total += num;
        if (total % 2 != 0) return false;
        int target = total / 2;

        boolean[] prev = new boolean[target + 1];
        boolean[] curr = new boolean[target + 1];
        prev[0] = true;
        curr[0] = true;

        for (int i = nums.length - 1; i >= 0; i--) {
            for (int t = 0; t <= target; t++) {
                boolean include = (t - nums[i] >= 0) ? prev[t - nums[i]] : false;
                boolean exclude = prev[t];
                curr[t] = include || exclude;
            }
            prev = curr.clone();
        }

        return prev[target];
    }
}
